man/man1/ttybiff.1
.TH TTYBIFF 1
.\" @(#) $Header: /xtel/pp/pp-beta/man/man1/RCS/ttybiff.1,v 6.0 1991/12/18 20:43:41 jpo Rel $
.\"
.\" $Log: ttybiff.1,v $
.\" Revision 6.0  1991/12/18  20:43:41  jpo
.\" Release 6.0
.\"
.\"
.\"
.SH NAME
ttybiff \- display new mail alerts
.SH SYNOPSIS
.in +.5i
.ti -.5i
ttybiff \%[-f\ file] \%[-u\ user] \%[-p\ number] \%[-s]
.in -.5i
.SH DESCRIPTION
This program is subsimed by xalert and will probably disappear from
future releases.
.PP
Ttybiff is a simple program designed to display new message alerts on
a users terminal. It does very little else, relying on its client to
do all the hard work. It is normally run in the background. it
displays the message on all terminals the user is logged onto on the
current machine, or if that fails to the standard output.
.PP
It has the following options
.TP
\-f file set the alert filename that \fIttybiff\fP and
\fIflagmail\fP communicate via. This defaults to the file \fB.alert\fP
in your home directory. This file is created by \fIttybiff\fP and read
by \fIflagmail\fP
.TP
\-u username
Allows you to pretend to be the given user. This does not really do
much as you will generally only get notified of messages to you, but
it may be useful.
.TP
\-p number
Change the default udp port that \fIttybiff\fP and \fIflagmail\fP
communicate via (normally assigned by the system).
.TP
\-s
Only display on the standard output, not on all terminals that the
user is logged into.
.SH FILES
.ta \w'~/.alert\0\0'u
~/.alert	The default alert file.
.SH "SEE ALSO"
flagmail(1), ttyalert(1)
.br
\fIPP Manual \- Volume 3: User Manual\fP
.SH "BUGS"


CHANGES









             Changes from PP version 5.2 to 6.0





_1.  _O_v_e_r_v_i_e_w

This  short  document  is  a   description   of   the   more
user/administrator  visible changes that have taken place in
PP between the last interim version (5.2) and the beta  ver-
sion  (6.0). This document is not exhaustive or complete but
is a guide to what to look out for. In general the following
may be assumed.

o+    The Queue files are compatible  between  5.2  and  6.0.
     Note  however that the Queue structure has changed. Old
     channels will have to be recompiled  to  recognise  the
     new format, but new channels will read old queue files.

o+    The qmgr protocol is compatible between 5.2 and 6.0

o+    The syntax of some of  the  table  files  has  changed.
     Where appropriate conversion scripts are provided.

o+    Some procedure calls have been  changed  or  have  dif-
     ferent parameters.

o+    the libdsap.a library is now required for compilation.

In summary, you should be careful when upgrading to the  6.0
release.   It  will  not  work  by  just  replacing  the 5.2
binaries with 6.0. For definite the domain table  will  HAVE
to be changed.

_2.  _D_e_t_a_i_l_e_d _c_h_a_n_g_e_s

This list is in no particular order of preference or  impor-
tance.

Domain Table
     The domain table syntax has been  changed  to  be  more
     explicit  and  has  been  brought into line with the OR
     table syntax. We feel this is clearer and more logical.
     Tools  that  process  the domain table have been taught
     the   new   format   and   there   are    scripts    in
     Tools/tables/compat  to convert between the old and new
     formats.  There is NO backwards compatibility here. Old
     format tables will just plain not work.

Aliases Table



                     December 18, 1991





                           - 2 -


     The aliases table syntax has changed to a  main  choice
     of  synonym or alias with further qualifiers specifying
     the address type of the synonym or alias.  The  changes
     are

          x:alias y         x:alias y
          x:synonym y       x:synonym y
          x:822-synonym y   x:synonym y 822
          x:x400-synonym y  x:synonym y x400
          x:822 y           x:alias y 822
          x:x400 y          x:alias y x400
     where any white space in  y  must  be  quoted  using  a
     backslash  There  is also a "external" qualifier.  This
     disables further lookups of synonyms  and  aliases  and
     can  be  used  for  a  main  site mail hub mapping from
     site_wide_name@localsite                             to
     local_machine_name@localmachine  and  vice  versa.  The
     use of external stops the looping alias  problem  which
     would otherwise occur.

Return of Contents
     Return of contents is now supported.  This  is  a  side
     effect  and  part  of  the  work on routing of delivery
     reports.

Use of pepsy
     Use of the PEPSY ASN.1 compiler rather  than  PEPY/POSY
     is  now  used  in  many  places. This makes for smaller
     binaries and much quicker compilation.

ISODE 7.0
     PP 6.0 is aligned to ISODE 7.0. It will not work  under
     ISODE 6.0.  If you are using the 6.0 release you should
     be using ISODE 7.0. It should however work under  ISODE
     6.8.

Tailor file
     There have been many changes to the  tailor  file.  The
     introduction of several new variables and the splitting
     up of header and body part types. These are

          dap_user     Directory username for DS lookups
          dap_passwd   Directory Passwd for same
          fsync        disable use of fsync(2)
          headertype   list of headers c.f. bodypart
          queuestruct  use hierarchical queue structure
          submit_addr  call submit over network
          diskuse      limit the disk space consumed by PP

     Channel tailoring also has a number of new variables

          drchan       replaces old outchan
          hdrin        headers inbound
          hdrout       headers outbound



                     December 18, 1991





                           - 3 -


          ininfo       inbound channel info
          inadr        inbound channel addr type
          insubadr     inbound channel sub address type
          intable      inbound channel table
          key          give alternative names to channel
          mtatable     list of known mtas for channel binding
          outinfo      outbound channel info
          outsubadr    outbound channel sub address
          outtable     outbound channel table
          solo-proc    channel requires address one at a time
          bad-sender-policyset the policy for unroutable senders
          check        set the checking mode of the channel

Channel pairing
     Channels may now be explicitly  paired  together.  This
     allows one definition to specify both an inbound and an
     outbound channel.  It also allows several logical chan-
     nels  for  one  inbound protocol by selecting a channel
     per MTA.

Timezones
     Many more lettered timezones are recognised. This  list
     is ambiguous at best and numeric forms should really be
     used except at user interfaces.

Delivery Reports
     Delivery reports are now routed back to their  origina-
     tor.  In 5.2 they were returned directly down the chan-
     nel they arrived on. This allows for one way  links  in
     general  makes  things  easier. It also allows protocol
     conversion so that, for instance, a delivery report  on
     an  X.400 message may go back via an RFC-822 if that is
     the only option.  A side effect of this is the  channel
     variable  outchan  has been renamed drchan and has dif-
     ferent semantics attached to it (see the manual).

Header parts
     Separation of headers and bodyparts for  each  channel.
     This  is  mainly a syntactic change but will allow more

Uip/rcvalert/ttybiff.c
/* ttybiff.c: tty alert program using the UDP syntax stuff */

# ifndef lint
static char Rcsid[] = "@(#)$Header: /xtel/pp/pp-beta/Uip/rcvalert/RCS/ttybiff.c,v 6.0 1991/12/18 20:39:41 jpo Rel $";
# endif

/*
 * $Header: /xtel/pp/pp-beta/Uip/rcvalert/RCS/ttybiff.c,v 6.0 1991/12/18 20:39:41 jpo Rel $
 *
 * $Log: ttybiff.c,v $
 * Revision 6.0  1991/12/18  20:39:41  jpo
 * Release 6.0
 *
 */



#include <stdio.h>
#include <isode/manifest.h>
#include "sys.file.h"
#include <sys/stat.h>
#include <utmp.h>
#include <sys/time.h>
#include <pwd.h>
#include <ctype.h>

#include "data.h"

char	username[40];
char	hostname[128];
char	homedir[512];
char	*filename = ".alert";
short	port = 0;
char	*extra = "NEW: ";
char	*format = "%extra(%size) %20from %subject << %50body";
int	width = 80;
int	debug = 0;
int	stdoutonly = 0;

char	*myname;

extern char	*index ();

main (argc, argv)
int	argc;
char	*argv[];
{
	extern char	*optarg;
	extern int	optind;
	int	opt;
	int	sd;

	myname = argv[0];
	while((opt = getopt(argc, argv, "f:p:u:s")) != EOF)
		switch (opt) {
		    case 's':
			stdoutonly = 1;
			break;
		    case 'f':
			filename = optarg;
			break;
		    case 'p':
			port = atoi(optarg);
			break;
		    case 'u':
			(void) strcpy (username, optarg);
			break;
		    default:
			fprintf (stderr,
				 "Usage: %s [-f file] [-p port] [-u user]",
				 myname);
			break;
		}
	argc -= optind;
	argv += optind;

	init_stuff ();

	sd = udp_start (port, filename, homedir);

	for (;;) {
		fd_set rfds;
		char	buffer[BUFSIZ];
		char	from[BUFSIZ];
		int	n;

		FD_ZERO (&rfds);
		FD_SET (sd, &rfds);
		if(select (sd + 1, &rfds, NULLFD, NULLFD,
			   (struct timeval *)0) <= 0)
			continue;

		n = sizeof (buffer) - 1;
		if (FD_ISSET (sd, &rfds) &&
		    getdata (sd, buffer, &n, username, from)) {
			buffer[n] = 0;
			if (stdoutonly ||
			    (notify_normal (buffer, username) == NOTOK &&
			    isatty (1)))
				printf ("\n\r%s%c\n\r",
					buffer, '\007');
		}
	}
}

init_stuff ()
{
	struct passwd *pwd;
	char	*cp, *getenv ();

	if ((pwd = getpwuid (getuid())) == NULL) {
		fprintf (stderr, "No user name\n");
		exit(-1);
	}
	if (username[0] == NULL)
		(void) strcpy (username, pwd -> pw_name);

	if ((cp = getenv ("HOME")) != NULL && *cp)
		(void) strcpy (homedir, cp);
	else
		(void) strcpy (homedir, pwd -> pw_dir);
}
